src/rest_api/classes/laboratory/widgets/members.clj

```diff
@@ -1,31 +1,108 @@
+                                 :start_date (first from)
+                                 :end_date (first to) } }))
 
-(defn former-members [l]
-  {:data (some->> (:labloratory/past-lab-members l)
-                  (map get-lineage-data))
-   :description "former members of the laboratory"})
+
+(defn- generate-lineage [laboratory]
+  (let [db (d/entity-db laboratory)
+        reps (:laboratory/representative laboratory)
+        lineage
+        (apply merge
+               (flatten 
+                 (for [rep reps] 
+                   (->> (d/q '[:find ?other-person-id ?other-person-name (pull ?person-supervised-by-ent [*])
+                               :in $ ?person
+                               :where
+                               [?other-person-ent :person/supervised-by ?person-supervised-by-ent]
+                               [?person-supervised-by-ent :person.supervised-by/person ?person]
+                               [?other-person-ent :person/id ?other-person-id]
+                               [?other-person-ent :person/standard-name ?other-person-name]]
+                             db (:db/id rep))
+                        (map (fn [oid]
+                               (generate-map oid)))(seq)))))]
+    lineage))
+
+(defn current-members [laboratory]
+  (let [db (d/entity-db laboratory)
+        lineage (generate-lineage laboratory)
+        data (->> (d/q '[:find [?members ...]
+                         :in $ ?laboratory
+                         :where
+                         [?laboratory :laboratory/registered-lab-members ?members]]
+                       db (:db/id laboratory))
+                  (map (fn [members]
+                         (let [member (d/entity db members)
+                               lab (d/entity db laboratory)]
+                           {:name (pack-obj "person" member)
+                            :level (:level ((keyword (:person/id member)) lineage))
+                            :start_date (:start_date ((keyword (:person/id member)) lineage))
+                            :end_date (:end_date ((keyword (:person/id member)) lineage))
+                            :duration (:duration ((keyword (:person/id member)) lineage)) })))(seq))]
+    {:data data
+     :description "current members of the laboratory"}))
+
+(defn former-members [laboratory]
+  (let [db (d/entity-db laboratory)
+        lineage (generate-lineage laboratory)
+        data (->> (d/q '[:find [?members ...]
+                         :in $ ?laboratory
+                         :where
+                         [?laboratory :laboratory/past-lab-members ?members]]
+                       db (:db/id laboratory))
+                  (map (fn [members]
+                         (let [member (d/entity db members)
+                               lab (d/entity db laboratory)]
+                           {:name (pack-obj "person" member)
+                            :level (:level ((keyword (:person/id member)) lineage))
```

Comment:
This is same as above. You could pull ((keyword (:person/id member)) lineage)) into it's own variable

Reply:
I either don't understand the advantage that, or don't understand what you mean.  You mean in the let to assign that to its own variable ?  Would that be just for clarity, it seems like an extra variable.

Reply:
Changed.